File: backend/src/main/java/com/cinemaebooking/backend/showtime/domain/model/Showtime.java
package com.cinemaebooking.backend.showtime.domain.model;

import com.cinemaebooking.backend.common.domain.BaseEntity;
import com.cinemaebooking.backend.common.exception.domain.CommonExceptions;
import com.cinemaebooking.backend.showtime.domain.enums.Language;
import com.cinemaebooking.backend.showtime.domain.enums.ShowtimeStatus;
import com.cinemaebooking.backend.showtime.domain.valueobject.ShowtimeId;
import lombok.Getter;
import lombok.experimental.SuperBuilder;

import java.time.LocalDateTime;

@Getter
@SuperBuilder(toBuilder = true)
public class Showtime extends BaseEntity<ShowtimeId> {

    private Long movieId;
    private Long roomId;
    private Long roomLayoutId;
    private Long formatId;

    private LocalDateTime startTime;
    private LocalDateTime endTime;

    private Language audioLanguage;
    private Language subtitleLanguage;

    private ShowtimeStatus status;

    // =====================================================
    // BUSINESS METHODS
    // =====================================================

    public void update(LocalDateTime start,
                       LocalDateTime end) {

        validateStartEnd(start, end);

        this.startTime = start;
        this.endTime = end;
    }

    public void updateLanguage(Language audio, Language subtitle){
        validateLanguages(audio, subtitle);
        this.audioLanguage = audio;
        this.subtitleLanguage = subtitle;
    }
    public void changeStatus(ShowtimeStatus newStatus) {
        if (newStatus == null) {
            throw CommonExceptions.invalidInput("Showtime status cannot be null");
        }
        this.status = newStatus;
    }

    public void cancel() {
        this.status =ShowtimeStatus.CANCELLED;
    }

    public void validateForCreate() {
        validateMovieId(movieId);
        validateRoomId(roomId);
        validateFormatId(formatId);

        validateStartEnd(startTime, endTime);
        validateLanguages(audioLanguage, subtitleLanguage);

        if (status == null) {
            throw CommonExceptions.invalidInput("Showtime status cannot be null");
        }
        if (roomLayoutId == null || roomLayoutId <= 0) {
            throw CommonExceptions.invalidInput("roomLayoutId must be a positive number");
        }
    }

    // =====================================================
    // VALIDATION
    // =====================================================

    private void validateMovieId(Long id) {
        if (id == null || id <= 0) {
            throw CommonExceptions.invalidInput("movieId must be a positive number");
        }
    }

    private void validateRoomId(Long id) {
        if (id == null || id <= 0) {
            throw CommonExceptions.invalidInput("roomId must be a positive number");
        }
    }

    private void validateFormatId(Long id) {
        if (id == null || id <= 0) {
            throw CommonExceptions.invalidInput("formatId must be a positive number");
        }
    }

    private void validateStartEnd(LocalDateTime start, LocalDateTime end) {
        if (start == null) {
            throw CommonExceptions.invalidInput("startTime cannot be null");
        }
        if (end == null) {
            throw CommonExceptions.invalidInput("endTime cannot be null");
        }
        if (end.isBefore(start)) {
            throw CommonExceptions.invalidInput("endTime must be after startTime");
        }
    }

    private void validateLanguages(Language audio, Language subtitle) {
        if (audio == null) {
            throw CommonExceptions.invalidInput("Audio language cannot be empty");
        }
        if (subtitle == null) {
            throw CommonExceptions.invalidInput("Subtitle language cannot be empty");
        }
    }
}

File: frontend/src/types/showtime.ts
// ─── Showtime Format ──────────────────────────────────────────────────────────
// Seeded: 2D (extraPrice=0), 3D (extraPrice=30000), IMAX (extraPrice=60000)
export type FormatName = '2D' | '3D' | 'IMAX'

export interface ShowtimeFormatResponse {
  id:         number
  name:       FormatName
  extraPrice: number
}

// ─── Showtime ─────────────────────────────────────────────────────────────────
export type ShowtimeStatus = 'SCHEDULED' | 'ONGOING' | 'FINISHED' | 'CANCELLED'

export interface ShowtimeResponse {
  id:               number
  movieId:          number
  roomId:           number
  formatId:         number
  cinemaId:         number
  roomLayoutId:     number
  startTime:        string   // ISO 8601: "2026-05-10T14:00:00"
  endTime:          string
  audioLanguage:    string
  subtitleLanguage: string
  status:           ShowtimeStatus
}

export interface CreateShowtimeRequest {
  movieId:          number
  roomId:           number
  formatId:         number
  startTime:        string
  endTime:          string
  audioLanguage:    string
  subtitleLanguage: string
}

export interface UpdateShowtimeRequest {
  startTime:        string
  endTime:          string
  audioLanguage:    string
  subtitleLanguage: string
}